Design of OpenPipeline
================

<!-- README.md is generated by running 'quarto render README.qmd' -->

## Pipeline architecture overview

<div class="column-page">

<div id="fig-architecture">

<p>

``` mermaid

flowchart LR
  Raw[Raw\ndata]
  Dataset[Dataset]
  Ingestion[/Ingestion/]
  Preproc1[/Unimodal\nsingle-sample\nprocessing/]
  Preproc2[/Unimodal\nmulti-sample\nprocessing/]
  Integration[/Integration/]
  Interpretation[/Interpretation/]
  Conversion[/Conversion/]

  Raw --- Ingestion --> Split[/Split/] --> Preproc1 --> Concat[/Concat/] --> Preproc2 --> Merge[/Merge/] --> Integration --> Interpretation --> Conversion --> Dataset

  Ingestion -.-|"- Demux.\n- Mapping"| Ingestion
  Preproc1 -.-|"- Count QC\n- Doublet calling\n- Ambient RNA\n- Gatekeeper"| Preproc1
  Preproc2 -.-|"- Normalisation\n- Feature filtering\n- Feature selection\n- Batch correction?\n- Dim. red.?"| Preproc2
  Integration -.-|"- Dim. red.\n- Data integration"| Integration
  Interpretation -.-|"- Clustering\n- Cell typing\n- GRN\n- ..."| Interpretation

  linkStyle 10,11,12,13,14 stroke:#fff,stroke-width:0px,text-align:left;

  %% click Ingestion "./#ingestion"
  %% style Ingestion color:#2873c7,text-decoration:underline;
  %% click Preproc1 "./#single-sample-preproc"
  %% style Preproc1 color:#2873c7,text-decoration:underline;
  %% click Preproc2 "./#multi-sample-preproc"
  %% style Preproc2 color:#2873c7,text-decoration:underline;
  %% click Integration "./#integration"
  %% style Integration color:#2873c7,text-decoration:underline;
```

</p>

Figure 1: Overview single cell processing steps in OpenPipeline.
Rectangles are data objects, parallelograms are Viash modules or
subworkflows.

</div>

</div>

## Decisions (to make)

-   Try to avoid pipeline metromaps
    ([Example](https://github.com/nf-core/rnaseq/blob/master/docs/images/nf-core-rnaseq_metro_map_grey.png)).
    Instead, multiple pipelines using a different subset or ordering of
    the same components can be implemented to suit the needs of any
    particular requirement.

-   Division of responsibilities:

<div class="column-page">

<div id="fig-resp">

<p>

``` mermaid

flowchart LR
  Ingestion[/Ingestion/]
  Preproc1[/Unimodal\nsingle-sample\nprocessing/]
  Preproc2[/Unimodal\nmulti-sample\nprocessing/]
  Integration[/Integration/]
  Interpretation[/Interpretation/]
  Conversion[/Conversion/]

  Raw --- Ingestion --> Split[/Split/] --> Preproc1 --> Concat[/Concat/] --> Preproc2 --> Merge[/Merge/] --> Integration --> Interpretation --> Conversion --> Dataset

  Ingestion -.-|"CZ Biohub"| Ingestion
  Split -.-|"Janssen"| Split
  Preproc1 -.-|"Janssen"| Preproc1
  Concat -.-|"Janssen"| Concat
  Preproc2 -.-|"Janssen"| Preproc2
  Merge -.-|"Janssen"| Merge
  Integration -.-|"Helmholtz"| Integration
  Interpretation -.-|"All\n(depends on\nspecific needs)"| Interpretation
  Conversion -.-|"Janssen"| Conversion

  linkStyle 10,11,12,13,14,15,16,17,18 stroke:#fff,stroke-width:0px;
```

</p>

Figure 2: Division of responsibilities.

</div>

</div>

<!-- conversion pipeline? -->

## Use cases

<div>

> **Note**
>
> In these use-cases, `Interpretation`, `Conversion` and `Dataset` are
> omitted because they are the same in every use case.

</div>

### A single unimodal sample

<div class="column-page">

<div id="fig-example1">

<p>

``` mermaid

flowchart LR

  Raw1[Sample 1] --- Ingestion1[/Ingestion/] --> Split1[/Split/] --> GEX1[GEX 1]
  GEX1 --> ProcGEX1[/USS GEX/]
  ProcGEX1 --- ConcatGEX[/Concat/] --> GEX[Combined\nGEX]
  GEX --> ProcGEX[/UMS GEX/]
  ProcGEX --- Merge[/Merge/] --> Integration[/Integration/] --> Downstream[/.../]

```

</p>

Figure 3: Example of how the concat and merges work.  
GEX: Gene-expression. USS: Unimodal single-sample processing. UMS:
Unimodal multi-sample processing.

</div>

</div>

### A single multimodal sample

<div class="column-page">

<div id="fig-example2">

<p>

``` mermaid

flowchart LR

  Raw1[Sample 1] --- Ingestion1[/Ingestion/] --> Split1[/Split/] --> GEX1[GEX 1] & ADT1[ADT 1] & RNAV1[RNAV 1]
  GEX1 --> ProcGEX1[/USS GEX/]
  ADT1 --> ProcADT1[/USS ADT/]
  RNAV1 --> ProcRNAV1[/USS RNAV/]
  ProcGEX1 --- ConcatGEX[/Concat/] --> GEX[Combined\nGEX]
  ProcADT1 --- ConcatADT[/Concat/] --> ADT[Combined\nADT]
  ProcRNAV1 --- ConcatRNAV[/Concat/] --> RNAV[Combined\nRNAV]
  GEX --> ProcGEX[/UMS GEX/]
  ADT --> ProcADT[/UMS ADT/]
  RNAV --> ProcRNAV[/UMS RNAV/]
  ProcGEX & ProcADT & ProcRNAV--- Merge[/Merge/] --> Integration[/Integration/] --> Downstream[/.../]

```

</p>

Figure 4: Example of how the concat and merges work.  
GEX: Gene-expression. ADT: Antibody-Derived Tags. RNAV: RNA Velocity.
USS: Unimodal single-sample processing. UMS: Unimodal multi-sample
processing.

</div>

</div>

### Multiple unimodal samples

<div class="column-page">

<div id="fig-example3">

<p>

``` mermaid

flowchart LR

  Raw1[Sample 1] --- Ingestion1[/Ingestion/] --> Split1[/Split/] --> GEX1[GEX 1]
  Raw2[Sample 2] --- Ingestion2[/Ingestion/] --> Split2[/Split/] --> GEX2[GEX 2]
  Raw3[Sample 3] --- Ingestion3[/Ingestion/] --> Split3[/Split/] --> GEX3[GEX 3]
  GEX1 --> ProcGEX1[/USS GEX/]
  GEX2 --> ProcGEX2[/USS GEX/]
  GEX3 --> ProcGEX3[/USS GEX/]
  ProcGEX1 & ProcGEX2 & ProcGEX3 --- ConcatGEX[/Concat/] --> GEX[Combined\nGEX]
  GEX --> ProcGEX[/UMS GEX/]
  ProcGEX --- Merge[/Merge/] --> Integration[/Integration/] --> Downstream[/.../]

```

</p>

Figure 5: Example of how the concat and merges work.  
GEX: Gene-expression. USS: Unimodal single-sample processing. UMS:
Unimodal multi-sample processing.

</div>

</div>

### Multiple multimodal samples

<div class="column-page">

<div id="fig-example4">

<p>

``` mermaid

flowchart LR

  Raw1[Sample 1] --- Ingestion1[/Ingestion/] --> Split1[/Split/] --> GEX1[GEX 1] & ADT1[ADT 1]
  Raw2[Sample 2] --- Ingestion2[/Ingestion/] --> Split2[/Split/] --> GEX2[GEX 2] & ADT2[ADT 2]
  GEX1 --> ProcGEX1[/USS GEX/]
  ADT1 --> ProcADT1[/USS ADT/]
  GEX2 --> ProcGEX2[/USS GEX/]
  ADT2 --> ProcADT2[/USS ADT/]
  ProcGEX1 & ProcGEX2 --- ConcatGEX[/Concat/] --> GEX[Combined\nGEX]
  ProcADT1 & ProcADT2 --- ConcatADT[/Concat/] --> ADT[Combined\nADT]
  GEX --> ProcGEX[/UMS GEX/]
  ADT --> ProcADT[/UMS ADT/]
  ProcGEX & ProcADT --- Merge[/Merge/] --> Integration[/Integration/] --> Downstream[/.../]

```

</p>

Figure 6: Example of how the concat and merges work.  
GEX: Gene-expression. ADT: Antibody-Derived Tags. USS: Unimodal
single-sample processing. UMS: Unimodal multi-sample processing.

</div>

</div>

## Subworkflows

### Ingestion

Purpose: Convert raw sequencing data or count tables into MuData data
for further processing.

<div class="column-page">

<div id="fig-ingestion">

<p>

``` mermaid
flowchart LR

  BCL["BCL<sup>*</sup>"]
  Fastq["Fastq<sup>*</sup>"]
  Ref["Reference"]
  RawDir["Raw out<sup>*</sup>"]
  RawCounts["Raw counts<sup>†</sup>"]

  Demux[/"Demux"/]
  Mapping[/"Mapping"/]
  Convert[/"Convert"/]

  BCL --- Demux --> Fastq
  Fastq & Ref --- Mapping --> RawDir --- Convert --> RawCounts


  BCL -.-|".tar.gz/Directory"| BCL
  Fastq -.-|".fastq.gz/.fastq"| Fastq
  Ref -.-|".tar.gz?"| Ref
  RawDir -.-|"Directory"| RawDir
  RawCounts -.-|".h5mu"| RawCounts

  linkStyle 7,8,9,10,11 stroke:#fff,stroke-width:0px,text-align:left;
```

</p>

Figure 7: Ingestion pipeline.  
`*`: Possible entry points.  
`†`: Output file(s)

</div>

</div>

### Single-sample unimodal processing

Purpose: Per modality fitering pipelines are available to select true
from false cells.

<div class="column-page">

<div id="fig-sspreproc">

<p>

``` mermaid
flowchart LR

  RawCounts["Raw counts<sup>*</sup>"]
  Processed["Processed\ncounts<sup>†</sup>"]

  Step1[/"Cell\nfiltering"/]
  Step2[/"Doublet\ncalling"/]
  Step3[/"Ambient RNA\ncorrection"/]
  
  RawCounts --> Step1 --> Step2 --> Step3 --> Processed
```

</p>

Figure 8: Single-sample processing pipeline.  
`*`: Possible entry points.  
`†`: Output file(s)

</div>

</div>

### Multi-sample unimodal processing

At the final stage of this part the different modalities are merged into
a single muon object. Merging can be performed by combining the
different filterings or only using a single filtering.

<div class="column-page">

<div id="fig-mspreproc">

<p>

``` mermaid
flowchart LR

  Processed["Processed\ncounts<sup>*</sup>"]
  Normalised["Normalised\ncounts<sup>†</sup>"]

  Step1[/"Feature filtering"/]
  Step2[/"Batch correction?"/]
  Step3[/"Normalisation"/]
  Step4[/"Feature selection"/]

  Processed --> Step1 --> Step2 --> Step3 --> Step4 --> Normalised
```

</p>

Figure 9: Multi-sample processing pipeline.  
`*`: Possible entry points.  
`†`: Output file(s)

</div>

</div>

### Integration

Purpose: Performs an integration pipeline for single cell data based on
a single or multiple modalities.

<div class="column-page">

<div id="fig-integration">

<p>

``` mermaid
flowchart LR

  Normalised["Normalised\ncounts<sup>*</sup>"]
  Integrated["Integrated\ndata<sup>†</sup>"]

  Step1[/"Data integration"/]
  Step2[/"Dimensionality\nreduction"/]
  
  Normalised --> Step1 --> Step2 --> Integrated
```

</p>

Figure 10: Integration pipeline.  
`*`: Possible entry points.  
`†`: Output file(s)

</div>

</div>

### Interpretation

Purpose: Take different dataset annotations and combine them together
into a single enriched dataset. The idea is to have a diff_muon object,
i.e. a muon object containing the changes of the original object where
data from the diff_muon will be pushed to the original muon object.

<div>

> **Warning**
>
> This is what we did a time ago and it has the drawback that it could
> make everything very slow. So we need to be able to aggregate diffs
> before adding them to the final object.

</div>

<!--



---

--- 

---

## Version 2021-10-13


__Questions:__

1. What with metadata on the cells?
2. How do we define cell ids? At the moment we use the default as in scanpy BCODE-DIGIT-SAMPLEID
3. Bookkeeping
4. Where do we do Ab/lipid/vcf demultiplexing? (After the count step?)
5. Meta-cells
6. Count imputations
7. Compute environment awareness ==> GPUs/TPUs/...


## Ingestion



Pipelines list:

- seq_demultiplex: Converts BCL to read data
- [read_mapping](pipeline/read_mapping.md): Converts reads/UMIs to count matrices
- input_conversion: Converts count matrices from different formats into h5ad data objects containing a single modality and sample
- vdj_mapping: Converts reads to clonotypes

## Per sample processing




### Concat

Purpose: Combining different samples together over different modalities.

Pipeline list: 

- [concat](pipeline/concat.md): Takes multiple single sample muons and combines them together into a single multi-sample muon object.

### Integration


### Transformation

Purpose: Perform transformations on the initial multi-sample muon object to generate novel annotations onto the dataset for further analysis.

Pipeline list:

- [Clustering](pipeline/clustering.md)
- TI: Trajectory inference
- GRN: Gene regulatory network reconstruction
- Celltyping
- ...





### Reporting

Purpose: Provide standardized reporting on the final dataset.

Pipeline list: 

- qc: General QC report on the data.
- pseudobulk: Differential expression report/pipeline.
- compositional analysis: Describing and comparing the distributions of different annotations. 
- ...

__This can be very flexible, we now use a ipynb that we can add to the pipelines that gets generated with the resulting data. As such we can reuse some standardized notebooks. The difficulty here is the dependency managment in the notebook.__


### Data definition

![Data flow overview](figures/data-flow.png)

## Version 2021-09-29

![Overview single cell processing pipelines - version 20210919](figures/pipelines-target-p1.png)
_Overview single cell processing pipelines._

Remarks:
1. What with multi-modal data integration? 
2. RNA-velocity is starting at the wrong location since it is only a different mapping step.
3. Muon objects to be used for the multi-modal data.


-->
